Tests/AK/TestString.cpp

```diff
@@ -271,3 +271,15 @@ TEST_CASE(find)
     EXPECT_EQ(a.find('b', 4), Optional<size_t> { 6 });
     EXPECT_EQ(a.find('b', 9), Optional<size_t> {});
 }
+
+// NOTE: The tested behavior is required for String.prototype.replace and String.prototype.replaceAll.
```

Comment:
I don't love the fact that a generic component is going to have "unintuitive behaviour" for the sake of one of its users...
can LibJS implement this itself, perhaps?
Though I'm not entirely sure what "intuitive behaviour" would be in this case...

Reply:
JavaScript does have a pretty specific definition for this method:
https://tc39.es/ecma262/#sec-stringindexof

Specifically `If searchValue is the empty String and fromIndex ≤ len, return fromIndex` is the rule that applies here. It might make sense to just implement `StringIndexOf` according to the spec and then use that in JS's `replace` and `replaceAll`.

Reply:
That would be my first idea on how to do this, but the other problem remains that "what should AK do in this scenario" a.k.a. if what js does is unintuitive, what _is_ intuitive?

Reply:
Interestingly, other languages behave the same as JS. I was kind of assuming the STL would return `std::string::npos` here:
```c++
std::size_t pos = std::string("abc").find("", 2);
assert(pos == 2);
```

And Python:
```python
pos = 'abc'.find('', 2)
assert(pos == 2)
```

TBH I don't know what I consider intuitive here, but there's probably an argument for our `find` implementation to behave in-line with these.

Reply:
I'm not at all against having this function behave one way or another, I'm simply against implementing uintuitive behaviour in a generic library for the sake of a single user.
If this is the norm, then let's have it, and just not call it "unintuitive", or even mention javascript at all.

Reply:
Ya agreed - can just call the test case `TEST_CASE(find_empty_needle)` or something.

Reply:
Good thoughts, I removed the references to JavaScript (and wrote in the commit description that we want the behavior to be in line with what e.g. the C++ standard library does).